### DISCUSSION TITLE
Unable to infer attributes of constrained generic type parameter outside of the class

### DISCUSSION THREAD

**Discussion by rdong8** [microsoft/pyright#9507] [Q&A] [answered]
I was going to open an issue for this, but I realize that I might just be wrong about what to expect. It seems that Pyright is unable to infer the constraints from a generic type parameter when it is referred to outside of the class, unless you specify a concrete type for it. 

```python
from abc import ABC, abstractmethod


class Currency(ABC):
    @abstractmethod
    def to_int(self) -> int:
        return 0


class Account[BaseCurrency: Currency]:
    def __init__(self, balance: BaseCurrency):
        self.balance = balance

    def get_balance(self) -> int:
        return self.balance.to_int()  # Works as expected


def get_balance(account: Account) -> int:
    return account.balance.to_int()  # to_int is "Unknown" to pyright
```

This only works if you do `get_balance[BaseCurrency: Currency](account: Account[BaseCurrency]) -> int`. Is this the expected behavior?

**Comment by erictraut** [comment, marked as answer]
Pyright is working correctly here. The `Account` class is generic and accepts one type argument. In the function `get_balance`, you have annotated the `account` parameter with the type `Account`. Since you have omitted the type argument, type checkers will assume `Any` (or in the case of pyright, `Unknown`, which is an implicit `Any`). That means the instance variable `balance` has the type `Unknown`.

This behavior is all prescribed by the Python typing spec, and pyright is conforming to the spec.
